### PR TITLE
Add collect_set Spark aggregation function

### PR DESCRIPTION
**Pull request by zhli1142015**
Doc: https://docs.databricks.com/en/sql/language-manual/functions/collect_set.html
Code: 
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala#L39C16-L39C23
https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/collect.scala#L147C12-L147C22
There are 3 semantic difference from `set_agg`:
1. Null values are excluded.
```
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
val jsonStr = """{"txn":null}"""
val jsonStr1 = """{"txn":null}"""
val jsonStr2 = """{"txn":null}"""
val jsonStr3 = """{"txn":null}"""
val jsonSchema = StructType(Seq(StructField("txn",LongType,true)))
val df = spark.read.schema(jsonSchema).json(Seq(jsonStr, jsonStr1, jsonStr2, jsonStr3).toDS)
df.select(collect_set($"txn")).show
+----------------+
|collect_set(txn)  |
+----------------+
|              []          |
+----------------+
```
2. Nested Nulls are allowed.
```
import org.apache.spark.sql.types._
val jsonStr = """{"txn":{"appId":"txnId","version":0,"lastUpdated":null}}"""
val jsonStr1 = """{"txn":{"appId":"txnId","version":1,"lastUpdated":1}}"""
val jsonStr2 = """{"txn":{"appId":"txnId","version":0,"lastUpdated":null}}"""
val jsonStr3 = """{"txn":{"appId":"txnId","version":1,"lastUpdated":null}}"""
val jsonSchema = StructType(Seq(StructField("txn", StructType(Seq(StructField("appId",StringType,true),StructField("lastUpdated",LongType,true),StructField("version",LongType,true))),true)))
val df = spark.read.schema(jsonSchema).json(Seq(jsonStr, jsonStr1, jsonStr2, jsonStr3).toDS)
df.select(collect_set(col("txn"))).show(false)
+---------------------------------------------------+
|collect_set(txn)                                                     |
+---------------------------------------------------+
|[{txnId, 1, 1}, {txnId, null, 0}, {txnId, null, 1}]         |
+---------------------------------------------------+
```
3. Map type is not allowed.

Changes:
1. Move `SetBaseAggregate` and `SetAggAggregate` to lib folder.
2. Register `SetAggAggregate` with `ignoreNulls=true` (1), `throwOnNestedNulls=false` (2) for complex types.
3. Not register map type for this function.

### CODE REVIEW

**Comment by zhli1142015** [comment]
cc @rui-mo and @PHILO-HE , thanks.

File: velox/functions/lib/aggregates/SetAggregateBase.h [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+#include "velox/exec/Aggregate.h"
+#include "velox/exec/SetAccumulator.h"
+#include "velox/functions/lib/CheckNestedNulls.h"
+
+using namespace facebook::velox::aggregate::prestosql;
+
+namespace facebook::velox::functions::aggregate {
+
+template <typename T, bool ignoreNulls = false>
+class SetBaseAggregate : public exec::Aggregate {
```
Comment (rui-mo) [inline velox/functions/lib/aggregates/SetAggregateBase.h:27]: nit: could we align the class name `SetBaseAggregate` and the file name `SetAggregateBase.h`?

Comment (rui-mo) [inline velox/functions/lib/aggregates/SetAggregateBase.h:27]: > Nested Nulls are allowed.

I wonder how this difference is handled. It is also through the ignoreNulls param, and if so, could we document it?

**Review by rui-mo — COMMENTED** [review]
Thanks!

File: velox/functions/lib/aggregates/SetAggregateBase.h (hunk 1 above)
Comment (zhli1142015) [reply velox/functions/lib/aggregates/SetAggregateBase.h]: we set `throwOnNestedNulls` to false, add comments for them. Thanks.

File: velox/docs/functions/spark/aggregate.rst [hunk 2]
```diff
@@ -58,6 +58,28 @@ General Aggregate Functions
     Returns an array created from the input ``x`` elements. Ignores null
     inputs, and returns an empty array when all inputs are null.
 
+.. spark:function:: collect_set(x) -> array<[same as x]>
+
+    Returns an array consisting of all unique values from the input ``x`` elements. 
+    Null values are excluded, and returns an empty array when all inputs are null.
+    Example::
```
Comment (rui-mo) [inline velox/docs/functions/spark/aggregate.rst:65]: nit: add an empty line above Example to improve readability.

**Review by rui-mo — COMMENTED** [review]
In the PR description, we talked about the semantic differences between Presto and Spark. Could we also explain how we solve it in this PR? Thanks.

**Comment by zhli1142015** [comment]
> In the PR description, we talked about the semantic differences between Presto and Spark. Could we also explain how we solve it in this PR? Thanks.

Updated, thanks.

**Comment by PHILO-HE** [comment]
@liujiayi771, could you review this pr? Thanks!

**Comment by zhli1142015** [comment]
@rui-mo , do you have more comments? Thanks.

File: velox/docs/functions/spark/aggregate.rst [hunk 3]
```diff
@@ -58,6 +58,29 @@ General Aggregate Functions
     Returns an array created from the input ``x`` elements. Ignores null
     inputs, and returns an empty array when all inputs are null.
 
+.. spark:function:: collect_set(x) -> array<[same as x]>
+
+    Returns an array consisting of all unique values from the input ``x`` elements. 
+    Null values are excluded, and returns an empty array when all inputs are null.
```
Comment (rui-mo) [inline velox/docs/functions/spark/aggregate.rst:64]: Can we document the handling of nested nulls and map type?

File: velox/functions/sparksql/aggregates/CollectSetAggregate.cpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+          case TypeKind::VARCHAR:
+            return std::make_unique<SetAggAggregate<StringView, true>>(
+                resultType);
+          case TypeKind::ARRAY:
+            [[fallthrough]];
+          case TypeKind::ROW:
+            // Nested nulls are allowed by setting 'throwOnNestedNulls' as
+            // false.
+            return std::make_unique<SetAggAggregate<ComplexType, true>>(
+                resultType, false);
```
Comment (rui-mo) [inline velox/functions/sparksql/aggregates/CollectSetAggregate.cpp:83]: I notice the default value for throwOnNestedNulls is false. Could you confirm?

**Review by rui-mo — COMMENTED** [review]
Thanks.

File: velox/functions/sparksql/aggregates/CollectSetAggregate.cpp (hunk 4 above)
Comment (zhli1142015) [reply velox/functions/sparksql/aggregates/CollectSetAggregate.cpp:83]: Yes, default value is false.  When `SetAggAggregate` used in Presto, `throwOnNestedNulls` is set as `exec::isRawInput`.

**Comment by rui-mo** [comment]
@mbasmanova This PR makes some modifications in lib/aggregates and prestosql/aggregates. Would you like to take a review? Thanks.

File: velox/functions/sparksql/aggregates/CollectSetAggregate.cpp (hunk 4 above)
Comment (rui-mo) [reply velox/functions/sparksql/aggregates/CollectSetAggregate.cpp:83]: Got it. Thanks.

File: velox/docs/functions/spark/aggregate.rst [hunk 5]
```diff
@@ -58,6 +58,31 @@ General Aggregate Functions
...
+    Nested nulls are compared as values.
+    Map type input is not supported as it's not orderable type in Spark.
```
Comment (mbasmanova) [inline velox/docs/functions/spark/aggregate.rst:66]: I wonder why maps are not supported. I understand that maps are not orderable, but there is no ordering happening in this function.

File: velox/functions/lib/aggregates/SetBaseAggregate.h [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/exec/Aggregate.h"
+#include "velox/exec/SetAccumulator.h"
+#include "velox/functions/lib/CheckNestedNulls.h"
+
+using namespace facebook::velox::aggregate::prestosql;
```
Comment (mbasmanova) [inline velox/functions/lib/aggregates/SetBaseAggregate.h:22]: No 'using namespace' in header files please.

File: velox/functions/sparksql/aggregates/tests/CMakeLists.txt [hunk 7]
```diff
@@ -34,6 +35,7 @@ target_link_libraries(
   velox_exec_test_lib
   velox_vector_test_lib
   velox_functions_aggregates_test_lib
+  velox_functions_spark
```
Comment (mbasmanova) [inline velox/functions/sparksql/aggregates/tests/CMakeLists.txt:38]: Wondering why we do we need this dependency now?

**Review by mbasmanova — COMMENTED** [review]
@rui-mo Looks fine to me. I assume SetBaseAggregate was just moved. Let me know if that's not the case and there are some material changes to it.

**Comment by rui-mo** [comment]
@mbasmanova Thank you for the review. That's the case, SetBaseAggregate was just moved.

File: velox/functions/sparksql/aggregates/tests/CollectSetAggregateTest.cpp [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+#include "velox/functions/sparksql/Register.h"
+#include "velox/functions/sparksql/aggregates/Register.h"
+
+using namespace facebook::velox::functions::aggregate::test;
+
+namespace facebook::velox::functions::aggregate::sparksql::test {
+
+namespace {
+
+class CollectSetAggregateTest : public AggregationTestBase {
```
Comment (rui-mo) [inline velox/functions/sparksql/aggregates/tests/CollectSetAggregateTest.cpp:27]: Can we add some tests on NaN values? Perhaps also assert throw for map type. Thanks.

File: velox/docs/functions/spark/aggregate.rst (hunk 5 above)
Comment (zhli1142015) [reply velox/docs/functions/spark/aggregate.rst]: Sorry, map type should be not comparable in Spark.

File: velox/functions/sparksql/aggregates/tests/CMakeLists.txt (hunk 7 above)
Comment (zhli1142015) [reply velox/functions/sparksql/aggregates/tests/CMakeLists.txt:38]: We need to use `array_sort` function in new added tests to ensure results are comparable.

File: velox/functions/sparksql/aggregates/tests/CollectSetAggregateTest.cpp (hunk 8 above)
Comment (zhli1142015) [reply velox/functions/sparksql/aggregates/tests/CollectSetAggregateTest.cpp:27]: Added, thanks.

File: velox/docs/functions/spark/aggregate.rst (hunk 5 above)
Comment (rui-mo) [reply velox/docs/functions/spark/aggregate.rst]: Spark uses compare to conduct ordering, are they the same in Spark? Thanks.

Comment (zhli1142015) [reply velox/docs/functions/spark/aggregate.rst]: This function is implemented with `HashSet` in Spark, which requires key should be comparable.
This should be my typo, I mean "Map type input is not supported as it's not comparable type in Spark." and updated it.
I added test cases to make sure this function has same behavior with Spark, specifically for some special cases like Nan, null, nested null.

Comment (rui-mo) [reply velox/docs/functions/spark/aggregate.rst]: Got it. Thanks.

**Comment by zhli1142015** [comment]
@rui-mo , @mbasmanova , @kagamiori , do you have more comments?
Thanks.

**Review by rui-mo — COMMENTED** [review]
Thanks. I have no further comment.

**Comment by zhli1142015** [comment]
@mbasmanova  and @kagamiori , could you help review this PR? Thanks.

**Review by mbasmanova — APPROVED** [review]
Looks good to me. Thanks.

**Pull request closed without merging**